### ISSUE TITLE
Distributed build for nixpkgs itself.

### ISSUE THREAD

**Issue by junjihashimoto** [tweag/rules_nixpkgs#208]
**Is your feature request related to a problem? Please describe.**
I think current rules_nixpkgs analyzes the nix expression and builds the derivations serially. 

**Describe the solution you'd like**
Build independent derivations in parallel by bazel's features.

**Comment by benradf** [comment]
> I think current rules_nixpkgs analyzes the nix expression and builds the derivations serially.

Is this the case? I think Bazel will run the `nixpkgs_package` repository rules in parallel if it can:
![image](https://github.com/tweag/rules_nixpkgs/assets/104896700/240f8c56-2670-4f2a-80fd-0a44f7b133bc)

Closing for now, but feel free to add more details if I've misunderstood your request.

**Comment by junjihashimoto** [comment]
@benradf When we set derivations for each target, it runs in parallel. However, in general, a derivations consists of some derivations. The derivations are managed by nix, so they would not be built by Bazel in parallel.

**Comment by junjihashimoto** [comment]
I think that if one derivation can be made into multiple bazel targets, the performance will improve dramatically.

**Comment by benradf** [comment]
> However, in general, a derivation consists of some derivations. The derivations are managed by nix, so they would not be built by Bazel in parallel.

I see, that makes sense. Thanks for clarifying.

We could pass [`--max-jobs auto`](https://nixos.org/manual/nix/stable/command-ref/nix-build.html#opt-max-jobs) to `nix-build` to make Nix build the derivations in parallel when Bazel invokes it. The problem with this approach is there is no coordination between Nix and Bazel as to how many threads to use, so we may over-saturate the processor with jobs and make the build slower instead of faster.

> I think that if one derivation can be made into multiple bazel targets, the performance will improve dramatically.

That would be a great feature to have, but I think it'll be quite difficult to achieve in general. We would have to somehow parse the top-level derivation and split it into Nix expressions for each component. Given the variety of derivations we might encounter, this won't always be possible.

If Bazel gets support for [Cooperative Parallelism](https://github.com/bazelbuild/bazel/issues/10443), we might be able to leverage that to coordinate thread usage efficiently between Nix and Bazel.
